pilot/pkg/config/coredatamodel/syntheticserviceentrycontroller.go

```diff
@@ -0,0 +1,375 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/pkg/log"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/mcp/sink"
+)
+
+var (
+	endpointKey = "networking.istio.io/endpointsVersion"
+	serviceKey  = "networking.istio.io/serviceVersion"
+)
+
+// SyntheticServiceEntryController is a temporary storage for the changes received
+// via MCP server
+type SyntheticServiceEntryController struct {
+	// keys [namespace][name]
+	configStore   map[string]map[string]*model.Config
+	configStoreMu sync.RWMutex
+	options       *Options
+	synced        uint32
+
+	discovery Discovery
+}
+
+// NewSyntheticServiceEntryController provides a new incremental CoreDataModel controller
+func NewSyntheticServiceEntryController(options *Options, d Discovery) CoreDataModel {
+	return &SyntheticServiceEntryController{
+		configStore: make(map[string]map[string]*model.Config),
+		options:     options,
+		discovery:   d,
+	}
+}
+
+// ConfigDescriptor returns all the ConfigDescriptors that this
+// controller is responsible for
+func (c *SyntheticServiceEntryController) ConfigDescriptor() model.ConfigDescriptor {
+	return model.ConfigDescriptor{model.SyntheticServiceEntry}
+}
+
+// List returns all the SyntheticServiceEntries that is stored by type and namespace
+// if namespace is empty string it returns config for all the namespaces
+func (c *SyntheticServiceEntryController) List(typ, namespace string) (out []model.Config, err error) {
+	if typ != model.SyntheticServiceEntry.Type {
+		return nil, fmt.Errorf("list unknown type %s", typ)
+	}
+
+	c.configStoreMu.Lock()
+	defer c.configStoreMu.Unlock()
+
+	// retrieve all configs
+	if namespace == "" {
+		for _, allNamespaces := range c.configStore {
+			for _, config := range allNamespaces {
+				out = append(out, *config)
+			}
+		}
+		return out, nil
+	}
+
+	byNamespace, ok := c.configStore[namespace]
+	if !ok {
+		return nil, nil
+	}
+
+	for _, config := range byNamespace {
+		out = append(out, *config)
+	}
+	return out, nil
+}
+
+// Apply receives changes from MCP server and creates the
+// corresponding config
+func (c *SyntheticServiceEntryController) Apply(change *sink.Change) error {
+	if change.Collection != model.SyntheticServiceEntry.Collection {
+		return fmt.Errorf("apply type not supported %s", change.Collection)
+	}
+
+	defer atomic.AddUint32(&c.synced, 1)
+
+	if change.Incremental {
+		// removed first
+		c.removeConfig(change.Removed)
```

Comment:
That makes sense, however my intention behind having individual locks was to reduce the locking time, since not all the operation inside the `incrementalUpdate` requires locking. Every time `IncrementalUpdate` receives resources it loops over them and coverts them to config objects, my assumption was excluding this can improve lock contention slightly.

Reply:
I suspect that if we don't call out (e.g. to do EDS updates) while we have the lock, we should be fine.